sklearn/linear_model/ridge.py

```diff
@@ -1087,55 +1310,54 @@ def fit(self, X, y, sample_weight=None):
             X, y, self.fit_intercept, self.normalize, self.copy_X,
             sample_weight=sample_weight)
 
-        gcv_mode = self.gcv_mode
-        with_sw = len(np.shape(sample_weight))
-
-        if gcv_mode is None or gcv_mode == 'auto':
-            if sparse.issparse(X) or n_features > n_samples or with_sw:
-                gcv_mode = 'eigen'
-            else:
-                gcv_mode = 'svd'
-        elif gcv_mode == "svd" and with_sw:
-            # FIXME non-uniform sample weights not yet supported
-            warnings.warn("non-uniform sample weights unsupported for svd, "
-                          "forcing usage of eigen")
-            gcv_mode = 'eigen'
+        gcv_mode = _check_gcv_mode(X, self.gcv_mode)
 
         if gcv_mode == 'eigen':
             _pre_compute = self._pre_compute
             _errors = self._errors
             _values = self._values
         elif gcv_mode == 'svd':
             # assert n_samples >= n_features
-            _pre_compute = self._pre_compute_svd
-            _errors = self._errors_svd
-            _values = self._values_svd
-        else:
-            raise ValueError('bad gcv_mode "%s"' % gcv_mode)
+            if sparse.issparse(X):
+                _pre_compute = self._pre_compute_svd_sparse
+                _errors = self._errors_svd_sparse
+                _values = self._values_svd_sparse
+            else:
+                _pre_compute = self._pre_compute_svd_dense
+                _errors = self._errors_svd_dense
+                _values = self._values_svd_dense
 
         if sample_weight is not None:
             X, y = _rescale_data(X, y, sample_weight)
-
-        centered_kernel = not sparse.issparse(X) and self.fit_intercept
-
-        v, Q, QT_y = _pre_compute(X, y, centered_kernel)
+            self._sqrt_sw = np.sqrt(sample_weight)
+            self._normalized_sqrt_sw = self._sqrt_sw / np.linalg.norm(
+                self._sqrt_sw)
+            self._weight_sum = sample_weight.sum()
+            self._with_sw = True
+        else:
+            self._with_sw = False
+            self._sqrt_sw = np.ones(X.shape[0], dtype=X.dtype)
+            self._weight_sum = X.shape[0]
+        self._sqrt_sw_matrix = sparse.dia_matrix(
+            (self._sqrt_sw, 0), shape=(X.shape[0], X.shape[0]))
+        precomputed = _pre_compute(X, y)
```

Comment:
This can be computed right before `for i, alpha in enumerate(self.alphas):`, and using `v, Q, QT_y` originally is a little more clearer.

Reply:
thanks!
I used a tuple instead of the individual variable names because the intermediary
variables computed by `_precompute_svd_sparse` are different, so it would be
strange to give them the same names. I still kept this precompute /
errors_and_values_helper organization in order to avoid changing the structure
of the code too much. another option would be to have pre_compute return a
dictionary, or have errors_and_values call pre_compute?